src/steps/extract-metadata.js

```diff
@@ -293,10 +293,16 @@ export default function extractMetaData(state, req) {
     }
   }
 
-  // remove undefined metadata
   for (const name of Object.keys(metadata)) {
     if (metadata[name] === undefined) {
+      // remove undefined metadata
       delete metadata[name];
+    } else if (Array.isArray(metadata[name])) {
+      // make absolute URLs for arrays
+      metadata[name] = metadata[name].map((value) => makeAbsoluteURLForMeta(state, value));
```

Comment:
We could reuse [`rewriteUrls`](https://github.com/adobe/helix-html-pipeline/blob/main/src/steps/utils.js#L184) here. That util is also used in the `rewrite-urls` step and turns URLs like this into path-only URLs. Or do we really need absolute URLs here?

Reply:
> Or do we really need absolute URLs here?

or add a flag `makeAbsolute` to  `rewrite-urls`()

Reply:
also, it would probably be better to apply the rewriting in render, as the logic in metadata is already a bit complex:
https://github.com/adobe/helix-html-pipeline/blob/c7b98c41f39ae3ede80f1db38d91e550543b0d98/src/steps/render.js#L62-L72

Reply:
meta are strings only (not anchor with href), I think it must stay absolute urls - you would need a special treatment to consume the meta if it can be a relative or an absolute url (if external link). In fact, making them relative might break all existing logic...

I initially started with `rewriteUrls` but several things:

1. it does not make absolute urls but relative - the extra `makeAbsolute` flag would be required and this adds extra logic which is currently isolated in "my" `makeAbsoluteURLForMeta` function.
2. it does not deal with branches - if an author page a url with `custombranch--repo--owner.hlx.page`, it is not considered - this could be an improvement of the current method
3. because `rewriteUrls` is used in different places, I did not want to interfere and take the risk to change some behaviour

I can re-try with `rewriteUrls` and see the impact.

Reply:
> also, it would probably be better to apply the rewriting in render, as the logic in metadata is already a bit complex:
> 
> https://github.com/adobe/helix-html-pipeline/blob/c7b98c41f39ae3ede80f1db38d91e550543b0d98/src/steps/render.js#L62-L72

But I assume we do not get the values from the metadata sheet in here... no ?

Reply:
ah, I just check, `render` comes after. We are already doing some special filtering in `extract-metadata`, I think this belongs here, not in `render`.

Reply:
yeah, maybe you are correct. it's better to do it here; or even as separate step.

Reply:
> it does not deal with branches - if an author page a url with custombranch--repo--owner.hlx.page, it is not considered - this could be an improvement of the current method

yes, it is:
```
const HELIX_URL_REGEXP = /^https:\/\/(?!admin\.|www\.)[^.]+\.(aem|hlx3?)\.(live|page)\/?.*/;
```

Reply:
> 2. it does not deal with branches

if this is a requirement, we should improve `rewriteUrl()`

Reply:
in any case, it would be a BREAKING CHANGE, which is currently a bit tricky with the helix5 changes. maybe we can introduce it only as new helix5 feature?

Reply:
actually, I don't think we should change this behaviour here, in favour of https://github.com/adobe/helix-html-pipeline/issues/497. eventually metadata could be written like `{{url.host}}/solutions/experiments/index-experimentation-exp01`. 

(also, maybe the developer intentionally wants to use his dev branch for testing, so rewriting would be a problem)

Reply:
dev branch for testing can be easily achieved in the client side code consuming the metadata: it needs the path only anyway and use the current host. And it is already the case today: the code cannot assume it gets the correct host anyway if the content is managed by an author.
The only problem that this solves is authors copy/pasting urls from any of the 3 hosts in the metadata of their Word / gdoc documents, as we tell them to do. If they paste a hlx.page url, we should just make sure it does not appear in production. All the rest is anyway out of scope and does to change here.

Reply:
And I do not think this is a "BREAKING CHANGE", slight change in behaviour, yes, but you have a URL today, you still get a URL after this. Different host, yes but points to same "content" in prod.

Reply:
> extra logic which is currently isolated in "my" `makeAbsoluteURLForMeta` function.

Ok, switching to relative links may be too breaking, but isn't similar behavior alreasy in `makeCanonicalHtmlUrl` today? I wouldn really try to reuse existing rewriting logic as much as possible.

Reply:
Nope. `makeCanonicalHtmlUrl ` is doing something else (only appending `index.html` where needed).
We have a large collection of methods doing similar but different things. That's why it is hard to plug on top without breaking backward compatibility. 
Also `rewriteUrl` should be named `makeRelative` because this is really what it is doing.

Reply:
> And I do not think this is a "BREAKING CHANGE", 

yes it is. projects might expect .hlx.page or hlx.live urls, and might be not be able to deal with the production url.